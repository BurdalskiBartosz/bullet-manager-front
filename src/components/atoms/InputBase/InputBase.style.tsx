import styled, { css } from 'styled-components';
import IconButton from '../IconButton';

type tWrapperProps = {
	isError: boolean | undefined;
	fullWidth: boolean;
	styleForm: 'FORM' | 'BASE';
};

export const StyledWrapper = styled.div<tWrapperProps>`
	position: relative;
	display: flex;
	flex-direction: column;
	width: ${({ fullWidth }) => (fullWidth ? '100%' : '330px')};
	${({ styleForm }) => {
		switch (styleForm) {
			case 'FORM': {
				return css`
					${StyledInput} {
						padding: 0.5rem;
						border: none;
						background-color: ${({ theme }) => theme.colors.light};
						border-bottom: 1px solid
							${({ theme }) => theme.colors.lightGray};
						&::placeholder {
							color: transparent;
						}
					}
					label {
						position: absolute;
						top: -50%;

						z-index: 1;
						transition: 0.3s ease;
						transform: translate(5px, 5px);
						font-weight: 400;
						&:has(
								+ .helper-selector
									> .input-helper-selector:placeholder-shown
							) {
							transform: translate(5px, 20px);
						}
						&:has(
								+ .helper-selector
									> .input-helper-selector:focus
							) {
							transform: translate(5px, 5px);
						}
					}
				`;
			}
			case 'BASE': {
				return css`
					${StyledInput} {
						padding: 1.2rem 1.5rem;
						border: 1px solid
							${({ theme }) => theme.colors.lightGray};
						border-radius: ${({ theme }) =>
							theme.sizes.borderRadius};
					}
				`;
			}
		}
	}}
	${({ isError }) =>
		isError &&
		css`
			& label {
				color: ${({ theme }) => theme.colors.danger};
			}
			& input {
				border-color: ${({ theme }) => theme.colors.danger};
			}
		`};
`;
export const StyledInnerWrapper = styled.div`
	position: relative;
`;
export const StyledInput = styled.input`
	outline: none;
	font-family: ${({ theme }) => theme.fonts.primary};
	width: 100%;
	&:focus {
		border-color: ${({ theme }) => theme.colors.dark};
	}
	&::placeholder {
		color: ${({ theme }) => theme.colors.gray};
		font-style: italic;
		opacity: 0.6;
		font-size: 13px;
	}
`;

export const StyledIconButton = styled(IconButton)`
	position: absolute;
	top: 50%;
	right: 20px;
	transform: translateY(-50%);
	cursor: pointer;
	transition: 0.3s ease;
	&:hover {
		fill: ${({ theme }) => theme.colors.purple100};
	}
`;

export const StyledError = styled.p`
	align-self: flex-end;
`;
