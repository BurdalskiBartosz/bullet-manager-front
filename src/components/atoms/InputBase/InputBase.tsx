import { FC, ReactNode } from 'react';
import { useTranslation } from 'react-i18next';
import Font from '../Font';
import { tIconNames } from '../Icon/Icon';
import {
	StyledIconButton,
	StyledInnerWrapper,
	StyledInput,
	StyledWrapper
} from './InputBase.style';

export type tInputBase = {
	id: string;
	label: string;
	type?: string;
	register?: Function;
	fullWidth?: boolean;
	isDisabled?: boolean;
	as?: 'textarea';
	children?: ReactNode | ReactNode[];
	icon?: {
		iconName: tIconNames;
		fn: () => void;
	};
	error: {
		isError: boolean;
		errorMessage: string;
	};
};

const InputBase: FC<tInputBase> = ({
	id,
	label,
	type = 'text',
	register,
	error,
	isDisabled,
	as,
	icon,
	children,
	fullWidth = true
}) => {
	const { t } = useTranslation();

	return (
		<StyledWrapper isError={error.isError}>
			<Font htmlFor={id} variant="label">
				{label}
			</Font>
			<StyledInnerWrapper>
				<StyledInnerWrapper>
					{children ? (
						children
					) : (
						<>
							<StyledInput
								as={as}
								id={id}
								type={type}
								disabled={isDisabled}
								placeholder={t(`${label} placeholder`)}
								fullWidth={fullWidth}
								{...(register ? register(id) : {})}
							/>
							{icon && (
								<StyledIconButton
									iconName={icon.iconName}
									fn={icon.fn}
								/>
							)}
						</>
					)}
				</StyledInnerWrapper>
			</StyledInnerWrapper>
			<Font
				variant="inputError"
				style={{ alignSelf: 'flex-end', height: '10px' }}
			>
				{error.isError ? error.errorMessage : ''}
			</Font>
		</StyledWrapper>
	);
};

export default InputBase;
