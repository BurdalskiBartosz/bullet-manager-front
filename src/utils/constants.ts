export const baseURL: string = 'http://localhost:3001/api';

export const appRootRoute: string = 'app';
