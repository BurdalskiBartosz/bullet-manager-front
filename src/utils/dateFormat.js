import { format } from 'date-fns';

export const dateFormat = (value) => format(value, 'MM/dd/yyyy');
