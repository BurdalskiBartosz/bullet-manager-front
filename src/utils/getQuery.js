export const getQuery = (object) => {
	if (!object) return '';
	const queryStrings = Object.entries(object).map(([key, value]) => {
		const stringKey = encodeURIComponent(key);
		if (typeof value === 'object')
			return `${stringKey}=${JSON.stringify(value)}`;
		else return `${stringKey}=${encodeURIComponent(value)}`;
	});

	if (queryStrings.length === 0) return '';
	return '?' + queryStrings.join('&');
};
