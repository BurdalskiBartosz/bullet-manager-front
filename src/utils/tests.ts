import { render } from '@testing-library/react';
import Providers from '../providers';

const customRender = (ui: any) => render(ui, { wrapper: Providers });

export * from '@testing-library/react';

export { customRender as render };
