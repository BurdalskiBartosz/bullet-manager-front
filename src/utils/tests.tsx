import { render, RenderOptions } from '@testing-library/react';
import { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
import { PreloadedState } from '@reduxjs/toolkit';
import { AppStore, RootState, setupStore } from 'store/store';
import { ThemeProvider } from 'styled-components';
import { theme } from 'styles/theme';
import { BrowserRouter } from 'react-router-dom';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>;
	store?: AppStore;
}
const reduxStore = setupStore();
export const renderWithProviders = (
	ui: React.ReactElement,
	{
		preloadedState = {},
		// Automatically create a store instance if no store was passed in
		store = reduxStore,
		...renderOptions
	}: ExtendedRenderOptions = {}
) => {
	const Wrapper = ({ children }: PropsWithChildren<{}>) => {
		return (
			<BrowserRouter>
				<Provider store={store}>
					<ThemeProvider theme={theme}>{children}</ThemeProvider>
				</Provider>
			</BrowserRouter>
		);
	};

	// Return an object with the store and all of RTL's query functions
	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export { renderWithProviders as render };
export * from '@testing-library/react';
