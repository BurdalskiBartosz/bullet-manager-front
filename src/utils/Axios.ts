/* eslint-disable no-param-reassign */
import axios, { AxiosRequestConfig, AxiosError, AxiosResponse } from 'axios';
import { baseURL } from './constants';

const options: AxiosRequestConfig = {
	baseURL,
	timeout: 10000,
	headers: { 'Content-Type': 'application/json' },
	withCredentials: true
};

export const fetchProvider = axios.create(options);
const { interceptors } = fetchProvider;

const getRequestConfig = (config: AxiosRequestConfig) => {
	const token: string | null = window.localStorage.getItem('token');

	if (token) config.headers!.Authorization = `Bearer ${token}`;

	return config;
};

const getRequestError = (error: AxiosError) => Promise.reject(error);

const getResponse = (response: AxiosResponse) => response;

const getResponseError = (error: AxiosError) => {
	const errorResponse =
		error.response?.status === 400 ? error : error.response;
	return Promise.reject(errorResponse);
};

interceptors.request.use(getRequestConfig, getRequestError);
interceptors.response.use(getResponse, getResponseError);

export const instanceForMocks = axios.create(options);
