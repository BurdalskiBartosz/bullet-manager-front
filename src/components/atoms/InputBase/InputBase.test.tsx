import { fireEvent, render, screen } from 'utils/tests';
import InputBase from './InputBase';

const mockFn = jest.fn();

const props = {
	id: 'test',
	label: 'Label test',
	error: {
		isError: false,
		errorMessage: 'Error is visible'
	}
};

describe('InputBase', () => {
	it('should correctly', () => {
		render(<InputBase {...props} />);
		const label = screen.getByLabelText(/Label test/i);
		expect(label).toBeInTheDocument();
		expect(label).toMatchSnapshot();
	});

	it('should correctly with icon', () => {
		render(
			<InputBase
				{...props}
				icon={{
					iconName: 'close',
					fn: mockFn
				}}
			/>
		);
		const icon = screen.getByTestId('icon-component');
		expect(icon).toBeInTheDocument();
		fireEvent.click(icon);
		fireEvent.click(icon);
		expect(mockFn.mock.calls.length).toBe(2);
	});

	it('should correctly with isDisabled prop', () => {
		render(<InputBase {...props} isDisabled />);
		const input = screen.getByPlaceholderText(`${props.label} placeholder`);
		expect(input).toBeInTheDocument();
		expect(input).toBeDisabled();
	});

	it('should correctly show error message', () => {
		const { rerender } = render(<InputBase {...props} />);
		const invisibleError = screen.queryByText(/Error is visible/i);
		expect(invisibleError).toBeNull();

		rerender(
			<InputBase
				{...props}
				error={{
					isError: true,
					errorMessage: props.error.errorMessage
				}}
			/>
		);
		const visibleError = screen.getByText(/Error is visible/i);
		expect(visibleError).toBeInTheDocument();
	});
});
