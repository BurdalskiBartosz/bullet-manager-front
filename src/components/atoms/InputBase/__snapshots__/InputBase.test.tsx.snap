// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputBase should correctly 1`] = `
.c0 {
  padding: 1.2rem 1.5rem;
  border-radius: 15px;
  outline: none;
  border: 1px solid #B7B7B7;
  font-family: Rubik,sans-serif;
  width: 100%;
}

.c0:focus {
  border-color: #171717;
}

.c0::-webkit-input-placeholder {
  color: #707070;
  font-style: italic;
  opacity: 0.6;
  font-size: 13px;
}

.c0::-moz-placeholder {
  color: #707070;
  font-style: italic;
  opacity: 0.6;
  font-size: 13px;
}

.c0:-ms-input-placeholder {
  color: #707070;
  font-style: italic;
  opacity: 0.6;
  font-size: 13px;
}

.c0::placeholder {
  color: #707070;
  font-style: italic;
  opacity: 0.6;
  font-size: 13px;
}

<input
  class="c0"
  id="test"
  placeholder="Label test placeholder"
  type="text"
  value=""
/>
`;
